pkg/util/cgroup/manager/cgroup.go

```diff
@@ -292,7 +292,54 @@ func DropCacheWithTimeoutWithRelativePath(timeoutSecs int, absCgroupPath string,
 	_, err := exec.Command("bash", "-c", cmd).Output()
 
 	delta := time.Since(startTime).Seconds()
-	general.Infof("[DropCacheWithTimeoutWithRelativePath] it takes %v to do \"%s\" on cgroup: %s", delta, cmd, absCgroupPath)
+	general.Infof("[DropCacheWithTimeoutAndAbsCGPath] it takes %v to do \"%s\" on cgroup: %s", delta, cmd, absCgroupPath)
+
+	// if this command timeout, a none-nil error will be returned,
+	// but we should return error iff error returns without timeout
+	if err != nil && int(delta) < timeoutSecs {
+		return err
+	}
+
+	return nil
+}
+
+func SetExtraCGMemLimitWithTimeoutAndRelCGPath(ctx context.Context, relCgroupPath string, timeoutSecs int, nbytes int64) error {
+	memoryAbsCGPath := common.GetAbsCgroupPath(common.CgroupSubsysMemory, relCgroupPath)
+
+	err := SetExtraCGMemLimitWithTimeoutAndAbsCGPath(timeoutSecs, memoryAbsCGPath, nbytes)
+	_ = asyncworker.EmitAsyncedMetrics(ctx, metrics.ConvertMapToTags(map[string]string{
+		"relCgroupPath": relCgroupPath,
+		"succeeded":     fmt.Sprintf("%v", err == nil),
+	})...)
+	return err
+}
+
+func SetExtraCGMemLimitWithTimeoutAndAbsCGPath(timeoutSecs int, absCgroupPath string, nbytes int64) error {
+	if nbytes == 0 {
+		return fmt.Errorf("invalid memory limit nbytes: %d", nbytes)
+	}
+
+	startTime := time.Now()
+
+	var interfacePath string
+	if common.CheckCgroup2UnifiedMode() {
+		if nbytes == 0 {
+			general.Infof("[SetExtraCGMemLimitWithTimeoutAndAbsCGPath] skip drop cache on %s since nbytes is zero", absCgroupPath)
+			return nil
+		}
+		//cgv2
+		interfacePath = filepath.Join(absCgroupPath, "memory.max")
+	} else {
+		//cgv1
+		interfacePath = filepath.Join(absCgroupPath, "memory.limit_in_bytes")
+	}
+
+	cmd := fmt.Sprintf("timeout %d echo %d > %s", timeoutSecs, nbytes, interfacePath)
+
+	_, err := exec.Command("bash", "-c", cmd).Output()
```

Comment:
just curious, why don't we just use io library instead of a bash command?

Reply:
because we want to reuse TIMEOUT(1)  user commands to trigger timeout to prevent stucking

Reply:
get